### PR TITLE
49 implement dotsync2 protocol

### LINKED ISSUE #49: Implement /dot/sync/2 protocol

### PR DESCRIPTION
**Pull request by asmie**
Implementation of the /dev/sync/2 protocol. 
Current implementation contain mechanism to send request and receive response using protobuf objects.
Implementation based on the other request/response protocol.

### CODE REVIEW

File: core/src/network/sync2/protobuf/sync.proto [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    optional bytes justification = 6;
+
+    // Indicates whether the justification is empty
+    optional bool empty_justification = 7;
+
+    // Below are fields not present in the current /dot/sync/2 spec (https://spec.polkadot.network/#defn-msg-block-response)
+    // but are supported by substrate https://github.com/paritytech/substrate/blob/master/client/network/sync/src/schema/api.v1.proto
+    optional bytes justifications = 8;
+    repeated bytes indexed_body = 9;
+}
```
Comment (mariopil) [inline core/src/network/sync2/protobuf/sync.proto:74]: Missing new line at the end of file

File: core/src/network/sync2/protocol.cpp [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+    // send request
+    auto reader_writer = std::make_shared<libp2p::basic::ProtobufMessageReadWriter>(std::move(stream_res.value()));
+    auto write_res = co_await resumeInCallback<Result<size_t>>(
+            [&reader_writer, &request](auto&& func){
+                reader_writer->write(request, std::move(func));
+            });
+
+    if (!write_res) {
+        m_log->debug("Writing to a stream caused error: {}", write_res.error().message());
```
Comment (mariopil) [inline core/src/network/sync2/protocol.cpp:52]: I would use here error message instead of debug.

File: core/src/network/sync2/protocol.cpp [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    auto stream_res = co_await resumeInCallback<Result<std::shared_ptr<Stream>>>(
+            [&peerId, &host = m_host](auto&& func) {
+                host.newStream(
+                        peerId,
+                        protocol,
+                        std::move(func));
+            });
+
+    if (!stream_res) {
+        m_log->debug("Getting stream caused an error: {}", stream_res.error().message());
```
Comment (mariopil) [inline core/src/network/sync2/protocol.cpp:40]: I would use here error message instead of debug.

File: core/src/network/sync2/protocol.cpp [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    // read response
+    auto read_res = co_await resumeInCallback<Result<proto::BlockResponse>>(
+            [&reader_writer](auto&& func) {
+                reader_writer->read<proto::BlockResponse>(std::move(func));
+            });
+
+    if (!read_res) {
+        m_log->debug("Reading from stream caused error: {}", read_res.error().message());
```
Comment (mariopil) [inline core/src/network/sync2/protocol.cpp:63]: The same regarding this debug message.

File: core/src/network/sync2/protocol.cpp [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        co_return read_res.error();
+    }
+
+    co_return std::move(read_res.value());
+}
+
+} // namespace plc::core::network::sync2
```
Comment (mariopil) [inline core/src/network/sync2/protocol.cpp:70]: Missing new line

**Review by mariopil — APPROVED** [review]

**Review by kylegranger — APPROVED** [review]

**Pull request merged**